src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs

```diff
@@ -74,152 +130,146 @@ private static bool TryParseDateTimeOffset(ReadOnlySpan<byte> source, out DateTi
-            }
-
-            int month;
-            if (!TryGetNextTwoDigits(source.Slice(start: 5, length: 2), out month))
-            {
-                goto ReturnFalse;
-            }
-
-            if (source[7] != JsonConstants.Hyphen)
-            {
-                goto ReturnFalse;
+                parseData.Year = (int)(digit1 * 1000 + digit2 * 100 + digit3 * 10 + digit4);
             }
 
-            int day;
-            if (!TryGetNextTwoDigits(source.Slice(start: 8, length: 2), out day))
+            if (source[4] != JsonConstants.Hyphen
+                || !TryGetNextTwoDigits(source.Slice(start: 5, length: 2), ref parseData.Month)
+                || source[7] != JsonConstants.Hyphen
+                || !TryGetNextTwoDigits(source.Slice(start: 8, length: 2), ref parseData.Day))
             {
-                goto ReturnFalse;
-            }
-
-            // We now have YYYY-MM-DD
-            bytesConsumed = 10;
-
-            int hour = 0;
-            int minute = 0;
-            int second = 0;
-            int fraction = 0; // This value should never be greater than 9_999_999.
-            int offsetHours = 0;
-            int offsetMinutes = 0;
-            byte offsetToken = default;
-
-            if (source.Length < 11)
-            {
-                goto FinishedParsing;
+                return false;
             }
 
-            byte curByte = source[10];
-
-            if (curByte == JsonConstants.UtcOffsetToken || curByte == JsonConstants.Plus || curByte == JsonConstants.Hyphen)
-            {
-                goto ReturnFalse;
-            }
-            else if (curByte != JsonConstants.TimePrefix)
-            {
-                goto FinishedParsing;
-            }
+            // We now have YYYY-MM-DD [dateX]
+
+            Debug.Assert(source.Length >= 10);
+            if (source.Length == 10)
+            {
+                // Just a calendar date
+                return FinishParsing(ref parseData, out value, out kind);
+            }
+
+            // Parse the time of day
+            // ---------------------
+            //
+            // ISO 8601-1:2019 5.3.1.2b "Local time of day complete extended format"
+            //  [timeX]   = [“T”][hour][“:”][min][“:”][sec]
+            //  [hour]    = [hh] [00 - 23] (4.3.8a)
+            //  [minute]  = [mm] [00 - 59] (4.3.9a)
+            //  [sec]     = [ss] [00 - 59, 60 with a leap second] (4.3.10a)
+            //
+            // ISO 8601-1:2019 5.3.3 "UTC of day"
+            //  [timeX][“Z”]
+            //
+            // ISO 8601-1:2019 5.3.4.2 "Local time of day with the time shift between
+            // local time scale and UTC" (Extended format)
+            //
+            //  [shiftX] = [“+”|“-”][hour][“:”][min]
+            //
+            // Notes:
+            //
+            // "T" is optional per spec, but _only_ when times are used alone. In our
+            // case, we're reading out a complete date & time (5.4.2.1b).
```

Comment:
```suggestion
            // case, we're reading out a complete date & time so it's required (5.4.2.1b).
```

I actually read this as "since we're a complete time we don't need it", then got confused when it was used as a basis of length computation.

Reply:
I'll  add the explicit clause.